File: src/data/courses/dsa/graphs/grid-graph-patterns/number-of-provinces/code/python/number_of_provinces.py
# ============================================================
# Number of Provinces
# ============================================================
class Solution:
    def findCircleNum(self, isConnected) -> int:
        n, count = len(isConnected), 0
        visited = [False] * n
        def dfs(city):
            visited[city] = True
            for j in range(n):
                if isConnected[city][j] == 1 and not visited[j]: dfs(j)
        for i in range(n):
            if not visited[i]: count += 1; dfs(i)
        return count

File: src/data/courses/dsa/graphs/grid-graph-patterns/number-of-provinces/code/java/NumberOfProvinces.java
// ============================================================
// Number of Provinces
// ============================================================
class Solution {
    public int findCircleNum(int[][] isConnected) {
        int n = isConnected.length, count = 0;
        boolean[] visited = new boolean[n];
        for (int i = 0; i < n; i++)
            if (!visited[i]) { count++; dfs(i, isConnected, visited); }
        return count;
    }
    void dfs(int city, int[][] mat, boolean[] visited) {
        visited[city] = true;
        for (int j = 0; j < mat.length; j++)
            if (mat[city][j] == 1 && !visited[j]) dfs(j, mat, visited);
    }
}
